### PR TITLE
Mpoly

### PR DESCRIPTION
**Pull request by wbhart**
The test code is done and passing. I've run all the gcd tests (except gcd) out to way more iterations than what is there now and they all pass (which is more than I can say for the nmod_mat constructors).

The remaining functions are considered helper functions of gcd and will be tested properly once I figure out the exact interface I want in the SparsePoly module. They are tested by gcd for now.

The divides function probably still needs fixing, once we decide what we want.

There's no documentation yet, but I've added docstrings. I'll add docs when I come back to Nemo in a few months.

The gcd test takes a while (6-7s): not much I can do about that. Generic gcd is just really, really slow.

### CODE REVIEW

**Comment by thofma** [comment]
The new tests produce too many deprecation warnings on 0.6 on travis. He aborts because the log is too big.

**Comment by wbhart** [comment]
There are some things I can fix there, e.g. the string warnings and the
call syntax I think I didn't fix yet. Perhaps I also still have the old
PolynomialRing syntax in one or two places which I should fix.

If I find some time I'll go through and see if I can reduce the number of
warnings slightly.

Bill.

**Comment by thofma** [comment]
Don't worry. It passes for me locally on 0.6 and latest master. I would want to merge it.

**Comment by wbhart** [comment]
It's up to you. We can fix those issue a little later. From memory they
mainly affect fmpz_mpoly.jl, which is currently switched off anyway.

**Comment by wbhart** [comment]
By the way, did the nmod_mat.constructors test pass for you locally? It
does not pass for me ever.

File: src/generic/MPoly.jl [hunk 1]
```diff
@@ -1,2847 +1,2989 @@
...
-                        monomial_sub!(temp, 1, maxn, 1, b[w].exps, n[w] + 1 - i, N)
-                        monomial_sub!(Exps, vw, temp, 1, Qe[w], k[w], N)
-                        if nheapinsert!(H, I, xn, vw, Exps, N, w) # either chain or insert into heap
-                           viewc -= 1
-                        end
-                     else
-                        push!(I, nheap_t(i, k[w], w, 0))
-                        vw = Viewn[viewc]
-                        monomial_sub!(Exps, vw, maxn, 1, b[w].exps, n[w] + 1 - i, N)
-
```
Comment (thofma) [inline src/generic/MPoly.jl]: If you change this to `isassigned(b, 1)`, most of the deprecations warnings are gone and tests won't time out.

Comment (thofma) [inline src/generic/MPoly.jl]: If you change this to isassigned(b, 1), most of the deprecations warnings are gone and tests won't time out.

Comment (thofma) [inline src/generic/MPoly.jl]: Wrong line, sorry. There is one call to `isdefined(b, 1)` somewhere. If you change it to `isassigned(b, 1)`, most of the deprecations warnings are gone and tests won't time out.

**Comment by thofma** [comment]
For me all tests are passing, including `nmod_mat` stuff.

**Comment by wbhart** [comment]
isdefined occurs 5 times in Nemo. I fixed the one in MPoly.jl. Do you want
me to add a patch to fix them all?

**Comment by wbhart** [comment]
Maybe its just the julia-0.5 that I'm using. It definitely still fails for
me.

  Expression: R([Z2(1) Z2(1);Z2(1) Z2(1);Z2(1) Z2(1)])
    Expected: Nemo.ErrorConstrDimMismatch
      Thrown: BoundsError

Does it look familiar?

**Comment by thofma** [comment]
No, I don't see it locally.

**Pull request merged**